Ability to define Area for dev:di:info CLI command #38758

rostilos opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
PR that adds the ability to specify in `php bin/magento dev:di:info <class> <area>` command the desired area code. 

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#38758

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Run `php bin/magento dev:di:info <class> <area>`
2. In the console you can see information about di for the selected class under a given scope ( standard GLOBAL + defined in the command parameters )

### Questions or comments
Example: 
`app/code/Magento/Backend/etc/adminhtml/di.xml`
![2024-05-26_05-21](https://github.com/magento/magento2/assets/85498741/6f2ec97f-2e7b-47c4-8984-0529372af025)


Note : on the current 2.4. branch, an error will occur when outputting information. It is not related to the current task, see :  https://github.com/magento/magento2/issues/38740

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

rostilos committed:
Merge pull request #1 from magento/2.4-develop

upd

rostilos committed:
Merge branch 'magento:2.4-develop' into 2.4-develop

rostilos committed:
[issue #38758] Area code parameter for dev:di:info CLI command

rostilos committed:
[issue #38758] Area code parameter for dev:di:info CLI command ( fixes )

rostilos committed:
Merge branch 'magento:2.4-develop' into 2.4-develop

engcom-Hotel reviewed (changes requested):
Hello @rostilos,

Thanks for the contribution!

Please resolve the conflicts to process this PR further.

Thanks

engcom-Charlie commented:
Hi @rostilos,

Thank you for your contribution!

Please resolve the conflicts in order to proceed further on this PR. Till then closing this PR. Once done feel free to reopen or let us know.

Thank you!

Commit:
Merge remote-tracking branch 'refs/remotes/origin/2.4-develop' into fix-for-issue-38758

# Conflicts:
#	app/code/Magento/Developer/Console/Command/DiInfoCommand.php

rostilos commented:
> Hi @rostilos,
> 
> Thank you for your contribution!
> 
> Please resolve the conflicts in order to proceed further on this PR. Till then closing this PR. Once done feel free to reopen or let us know.
> 
> Thank you!

@engcom-Charlie Hi! I'd like to reopen the PR, and add appropriate edits on conflict resolution

engcom-Charlie commented:
> > Hi @rostilos,
> > Thank you for your contribution!
> > Please resolve the conflicts in order to proceed further on this PR. Till then closing this PR. Once done feel free to reopen or let us know.
> > Thank you!
> 
> @engcom-Charlie Hi! I'd like to reopen the PR, and add appropriate edits on conflict resolution

@rostilos, I am reopening it. Please resolve the conflicts so that we can work on it further.

rostilos commented:
@engcom-Charlie
Yes, please reopen it ( since apparently I don't have the proper permissions to do so ). Conflicts are already resolved in a branch in my forked repository.

engcom-Charlie commented:
Hi @rostilos,

Please close this PR so that we can reopen this PR. Also resolve the conflicts in this PR itself in order to proceed further.

Thank you!

rostilos commented:
@engcom-Charlie Closed duplicate DP

engcom-Charlie commented:
> @engcom-Charlie Closed duplicate DP

Thank you! Reopening this PR.

engcom-Charlie committed:
Merge branch '2.4-develop' into fix-for-issue-38758

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-38758

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/Developer/Console/Command/DiInfoCommand.php:
```diff
@@ -173,4 +194,40 @@ protected function execute(InputInterface $input, OutputInterface $output)
+     * @return void
+     * @throws \InvalidArgumentException
+     */
+    private function setDiArea($area)
```
Please specify the return type of the method:

for e.g.
```suggestion
    private function setDiArea($area): void
```

engcom-Hotel commented on app/code/Magento/Developer/Console/Command/DiInfoCommand.php:
```diff
@@ -173,4 +194,40 @@ protected function execute(InputInterface $input, OutputInterface $output)
+     * @param string $area
+     * @return bool
+     */
+    private function validateAreaCodeFromInput($area)
```
```suggestion
    private function validateAreaCodeFromInput($area): boolean
```

engcom-Hotel commented on app/code/Magento/Developer/Console/Command/DiInfoCommand.php:
```diff
@@ -26,17 +29,30 @@ class DiInfoCommand extends Command
     public function __construct(
-        Information $diInformation
+        Information $diInformation,
+        AreaList    $areaList
```
New Params in Constructor should be null and handle it with ObjectManager

engcom-Hotel reviewed (changes requested):
Hello @rostilos,

Thanks for the contribution!

Please take a look at the review comments below and cover these changes with an automated test according to [DOD](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/#definition-of-done-1).

Thanks

Commit:
issue-38758 added new constructor param as null and handle with Object Manager, codestyle fixes

Commit:
issue-38758 code refactoring ( dev:di:info )

rostilos commented:
@engcom-Hotel 
I've added the requested edits, as well as - PR to the repository with documentation. 
But I have questions about the tests. I have added private methods, as far as I understand, they should not be covered by unit tests. Could you please help with this point?

engcom-Hotel commented on app/code/Magento/Developer/Console/Command/DiInfoCommand.php:
```diff
@@ -173,4 +195,41 @@ protected function execute(InputInterface $input, OutputInterface $output)
+        if ($this->validateAreaCodeFromInput($area)) {
+            $objectManager = ObjectManager::getInstance();
+
+            $areaOmConfiguration = $objectManager
```
Consider injecting the ObjectManager and ConfigLoader dependencies via the constructor to avoid direct usage of the ObjectManager.

rostilos replied:
I added ObjectManager to the constructor, but I don't quite understand about ConfigLoader ( its instance is not used, I just need the class name )

rostilos replied:
Also, in case we add OM to the constructor in the current approach, we get a cyclic dependency, right?

engcom-Hotel commented on app/code/Magento/Developer/Console/Command/DiInfoCommand.php:
```diff
@@ -173,4 +195,41 @@ protected function execute(InputInterface $input, OutputInterface $output)
+     * @return void
+     * @throws \InvalidArgumentException
+     */
+    private function setDiArea($area): void
```
The parameter $area should have a type hint of string.

rostilos replied:
fixed

engcom-Hotel reviewed (changes requested):
Hello @rostilos,

Thanks for making the changes. I have added more review points. Please consider those as well.

And regarding your query on `private` methods:

> But I have questions about the tests. I have added private methods, as far as I understand, they should not be covered by unit tests. Could you please help with this point?

Yes, you are correct, private methods can't be used in unit tests. If you want to use it in unit tests, make them protected or public.

Thanks

Commit:
issue-38703 codestyle issues fix

Commit:
issue-38703 codestyle ( prettify )

Commit:
issue-38758 added test coverage ( integration tests )

rostilos commented:
@engcom-Hotel please check it ( also added integration tests )

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-38758

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on dev/tests/integration/testsuite/Magento/Developer/Console/Command/DiInfoCommandTest.php:
```diff
@@ -0,0 +1,141 @@
+<?php
+
```
Please add `declare(strict_types=1);` in accordance of [technical guidelines](https://developer.adobe.com/commerce/php/coding-standards/technical-guidelines/#1-basic-programming-principles)
```suggestion
declare(strict_types=1);
```

rostilos replied:
Thanks, I will make the required edits and correct the tests a little later on

engcom-Hotel reviewed (changes requested):
Hello @rostilos,

Thanks for making the changes!

Now the changes seems good to me except the one below. Please correct the same and also look into the failed tests.

Thanks

Commit:
issue-38758 codestyle fixes ( static tests )

Commit:
Merge remote-tracking branch 'origin/fix-for-issue-38758' into fix-for-issue-38758

rostilos commented:
@magento run Static Tests

rostilos commented:
@magento run all tests

Commit:
issue-38758 codestyle fixes ( strict_types and copyright )

rostilos commented:
@magento run all tests

rostilos commented:
@engcom-Hotel Made the requested edits and also corrected the errors concerning codestyle. Failed functional tests seems flacky to me

engcom-Bravo committed:
Merge branch '2.4-develop' into fix-for-issue-38758

engcom-Bravo commented:
@magento run all tests

engcom-Bravo commented:
Hi @rostilos,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**

- Run php bin/magento dev:di:info <class> <area>
- In the console you can see information about di for the selected class under a given scope ( standard GLOBAL + defined in the command parameters )

**Before:** :heavy_multiplication_x: 

<img width="1437" alt="Screenshot 2024-09-12 at 13 00 49" src="https://github.com/user-attachments/assets/25b76711-983c-443f-9e86-58bedeb44031">



**After:** :heavy_check_mark: 

<img width="1437" alt="Screenshot 2024-09-12 at 13 00 59" src="https://github.com/user-attachments/assets/fbe665a7-7cde-445a-bdaa-2d1c157257db">


Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks

engcom-Charlie committed:
Merge branch '2.4-develop' into fix-for-issue-38758

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
One of the Functional B2B test failures in recent 2 builds is not constant and remaining one is a known issue. Hence moving the PR to Merge in Progress.

Run 1: 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38759/bab2e8630a660e333f7c876842e29c14/Functional/allure-report-b2b/index.html

Run 2: 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38759/32dca44fb17cbbdb8e9c9d55883d4d4d/Functional/allure-report-b2b/index.html

**Known issue :**
StorefrontPrintListItemsFromRequisitionListTest: [#AC-12166](https://jira.corp.adobe.com/browse/AC-12166)
